### ISSUE TITLE
Save and open Included Files automatically when opening model with included files

### ISSUE THREAD

**Issue by EwoutH** [NetLogo/NetLogo#2070]
Currently, after saving and closing a NetLogo model with Included Source Files, those files aren't opened (added to the interface/info/code bar) when opening the model again. This adds extra manual actions on each close and open, and could confuse other users who don't see large parts of the code.

So it would be great if the files that were opened when saving, are stored, and are opened again when you open the model (in the example below, `init.nls` and `tick.nls`).

![Screenshot_50](https://user-images.githubusercontent.com/15776622/207616417-9dd08223-8ba1-4f52-9f9a-16619b044da1.png)

**Comment by ERSUCC** [comment]
Now that we are working on a new file format, I added an implementation of this enhancement to the file format PR #2285. Note that for this enhancement to be model-specific, the model will have to be saved in a dirty state in order for the open tabs to persist correctly.

**Comment by brandesNW** [comment]
Will the open files be saved if one of the open included files is saved?

**Comment by ERSUCC** [comment]
No, you can save an included file without saving the main model, since it is a separate location in the filesystem. Ideally saving an included file would cause the open files to be saved, but that would be impossible without saving the main model if the open files are to be model-specific.

**Comment by brandesNW** [comment]
I understand that nls files are saved independently of the model, and can be used in multiple models. My thought was that maybe users could be asked if they want to reopen the model with the same nls files open. This could potentially happen when 1) users save a model that has nls files  2) users close a model that has nls files  , 3) users save an nls file.

--
Aaron Brandes, Software Developer
Center for Connected Learning and Computer-Based Modeling


From: Isaac Brown ***@***.***>
Date: Thursday, September 5, 2024 at 2:45 PM
To: NetLogo/NetLogo ***@***.***>
Cc: Aaron Andre Brandes ***@***.***>, Comment ***@***.***>
Subject: Re: [NetLogo/NetLogo] Save and open Included Files automatically when opening model with included files (Issue #2070)

No, you can save an included file without saving the main model, since it is a separate location in the filesystem. Ideally saving an included file would cause the open files to be saved, but that would be impossible without saving the main model if the open files are to be model-specific.

**Comment by ERSUCC** [comment]
The issue is that for the open files to be model-specific, they would have to be stored in the model file, which can't be saved just by saving an nls file. Imagine that a user does an edit that they intend to be temporary to the model file, then saves an nls file. We wouldn't want to force a save of the entire model purely to save the open tabs, because that would overwrite the model file with their temporary changes. The only good way to fix this would be to have a multi-file format, which it seems like people don't want (but I am currently options discussing with Jason).

**Comment by brandesNW** [comment]
I agree that under the scenario you give the user might not want to save a list of nls files  that are open to  the model. What about a file menu option “Save model with list of open nls files”
Anyway keep this on a list of things to consider.


--
Aaron Brandes, Software Developer
Center for Connected Learning and Computer-Based Modeling


From: Isaac Brown ***@***.***>
Date: Friday, September 6, 2024 at 1:45 AM
To: NetLogo/NetLogo ***@***.***>
Cc: Aaron Andre Brandes ***@***.***>, Comment ***@***.***>
Subject: Re: [NetLogo/NetLogo] Save and open Included Files automatically when opening model with included files (Issue #2070)

The issue is that for the open files to be model-specific, they would have to be stored in the model file, which can't be saved just by saving an nls file. Imagine that a user does an edit that they intend to be temporary to the model file, then saves an nls file. We wouldn't want to force a save of the entire model purely to save the open tabs, because that would overwrite the model file with their temporary changes. The only good way to fix this would be to have a multi-file format, which it seems like people don't want (but I am currently options discussing with Jason).

**Comment by ERSUCC** [comment]
That could work temporarily, but ideally we would be able to keep track of model changes separately from the open tabs while still associating open tabs with a particular model. I'll leave it as is for now and we can decide what to do about it later.

**Comment by ERSUCC** [comment]
This was implemented a while ago and finalized in commit [f5efeda](https://github.com/NetLogo/NetLogo/commit/f5efeda5593177e61debf4ef08022d8f382d7987). FYI, it does not require the model to be resaved, the currently open temp files are tracked whenever a temporary code tab is opened or closed.
